### PR TITLE
New architecture: BPF

### LINKED ISSUE #838: Implement an eBPF decompiler/disassembler
Hey,

Recently in a CTF Competition (33C3) there was a challenge which required analysis of a SECCOMP BPF - a mechanism that allows to filter syscalls using a BPF,  BPF disassembly is simple, as you probably know the opcode set is very small and backward branches are not available (BPF programs always terminate). 

I used this debug tool called bpf_dbg i found in the kernel tree (http://lxr.free-electrons.com/source/tools/net/bpf_dbg.c) in the competition, it was quite awkward and I had to convert the BPF to a list of struct's format the get it to work properly, here is the actual decompile function:
http://lxr.free-electrons.com/source/tools/net/bpf_dbg.c#L216
I am confident this will be useful to many people. I would be glad to contribute the code myself, can I get a green light to start working on this? (Haven't worked on this project before so didn't just want to throw up a pull request on you guys).

### PR DESCRIPTION
**Pull request by david942j**
closes #838 

Berkeley Packet Filter
---
- [X] Supports both classic and extended BPF (`CS_MODE_BPF_CLASSIC, CS_MODE_BPF_EXTENDED`)
- [X] Fully tested
- [X] Python bindings

references:
- http://www.tcpdump.org/papers/bpf-usenix93.pdf
- https://www.kernel.org/doc/Documentation/networking/filter.txt
- https://github.com/iovisor/bpf-docs/blob/master/eBPF.md

### CODE REVIEW

**Comment by aquynh** [comment]
wow nice work! let me go thru the code & give some comments

File: cstool/cstool_bpf.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	cs_regs regs_read, regs_write;
+	uint8_t regs_read_count, regs_write_count;
+
+	// detail can be NULL on "data" instruction if SKIPDATA option is turned ON
+	if (ins->detail == NULL)
+		return;
+
+	bpf = &(ins->detail->bpf);
+
+	unsigned i;
```
Comment (aquynh) [inline cstool/cstool_bpf.c]: please put all variable declaration at the beginning of code block, but no in the middle.

File: arch/BPF/BPFInstPrinter.c [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+#include "BPFInstPrinter.h"
+#include "BPFMapping.h"
+
+static cs_bpf_op *expand_bpf_operands(cs_bpf *bpf)
+{
+	/* assert(bpf->op_count < 3); */
+	return &bpf->operands[bpf->op_count++];
+}
+
+static void push_op_reg(cs_bpf *bpf, bpf_op_type val, uint8_t ac_mode) {
```
Comment (aquynh) [inline arch/BPF/BPFInstPrinter.c]: please put the open bracket `{` in the next line, but not at the end of function line.
(here and also some functions below)

**Comment by aquynh** [comment]
the code looks very clean, you did an amazing job!

@HarDToBelieve @catenacyber this also updates your regression testsuite & fuzzer, please ack.

**Comment by aquynh** [comment]
Few more comments:

- please update https://github.com/aquynh/capstone/blob/next/bindings/python/Makefile#L73

- please fix warnings reported by MSVC: https://ci.appveyor.com/project/aquynh/capstone/builds/22434061

File: suite/fuzz/fuzz_disasm.c [hunk 3]
```diff
@@ -189,14 +189,24 @@ static struct platform platforms[] = {
         CS_MODE_BIG_ENDIAN,
         "tms320c64x"
     },
-#if CS_NEXT_VERSION >= 5
```
Comment (catenacyber) [inline suite/fuzz/fuzz_disasm.c]: I think the ifdef needs to remain so that both `next` and `master` branch can be fuzzed

**Comment by catenacyber** [comment]
Great ! 

Cheers @david942j 
PS : if you know about BPF, could you take a look at https://github.com/the-tcpdump-group/libpcap/pull/777

File: suite/fuzz/fuzz_disasm.c (hunk 3 above)
Comment (david942j) [reply suite/fuzz/fuzz_disasm.c]: Oh I found both master and next branch has the same `CS_NEXT_VERSION = 5` that's why I removed it.

Comment (catenacyber) [reply suite/fuzz/fuzz_disasm.c]: Ok thanks @david942j so your code is good

@aquynh how can a program know at compile time which architectures are supported by the current capstone ie branch `master` and branch `next` ?

**Comment by aquynh** [comment]
merged, thanks for this amazing work!

**Comment by aquynh** [comment]
@david942j, i had a commit to print out BPF for `cstool -v` at https://github.com/aquynh/capstone/commit/7ed49b360e500d005bf0572b47176d4f84d19b90

**Comment by david942j** [comment]
Oops I missed it, thanks!

**Pull request merged**